docs/proposals/build-execution-using-build-run.md

```diff
@@ -0,0 +1,76 @@
+---
+title: build-execution-using-build-run
+authors:
+  - "@sbose78"
+  - "@zhangtbj"
+status: design
+---
+
+# Build execution using the BuildRun API
+
+
+## About
+
+A `BuildRun` is an immutable CR/object that represents a single execution of the `Build`.
+
+The simplest `BuildRun` looks like this:
+```
+apiVersion: build.dev/v1alpha1
+kind: BuildRun
+metadata:
+  generateName: account-service-build-
+spec:
+  buildRef:
+    name: kaniko-golang-build
+```
+
+The `BuildRun` API also provides a way to override/specify  execution time information. 
+
+As an example, a `Build` configuration should not have to accurately specify the resource requirements - the information on resource requirements is valuable at build execution time which varies from environment to environment.
+
+## Defining a BuildRun
+
+The `BuildRun`, apart from triggering a build execution, also provides the user an opportunity to specify/override a subset of Build configuration which are likely to change at build execution time.
+
+### What is deterministic
+
+#### What are we building
+* The source code that's being built into an image.
+* The application binary that's being packaged into an image.
+
+The above information wraps the integrity of the code
+that we are building into an image.
+
+#### How are we building
+* The `BuildStrategy` being used to build the image.
+* Inputs associated with the `BuildStrategy` : Dockerfile, builder image, etc.
+* The runtime base image being used to build a lean image.
+
+The above information wraps the know-how that drives the process of converting source code into an image.
+
+### What is non-deterministic
+
+* The Image being pushed to
+    * Repository
+    * Tag
+* The Service Account used to execute the build.
+* Execution-time resource requiements.
+
+As an example,
+the source code https://github.com/nodeshift-starters/nodejs-rest-http-crud might be pushed to different image repositories depending on who is executing the build. 
+
+### Deciding what should be overriden
+
+* Could the value of attribute X be reasonably determined at the time of `Build` configuration specification ?
+* Could the modification of attribute X reasonably compromise the integrity of the build ?
+* Could the value of attribute X reasonably differ in the context of a clusters or a namespace ?
+
+The above is a non-exhaustive list of questions that should help us modify the BuildRun API in future.
+
+### Next steps / Consequences
+
+Here's what it means to adopt this proposal: 
+
+1. Make `spec.output` optional in `Build` API's `spec` .
+2. Introduce `spec.output` in the `BuildRun` API's `spec`.
```

Comment:
I would add as consequences that introducing `spec.output`  in the BuildRun can lead to use cases, where a change on the `spec.output` implies a change on the `spec.source`, but not having the `spec.source` in the `BuildRun` will be confusing. At the same time having the `spec.output` and the `spec.source` in the `BuildRun` already compromises the integrity of the `Build`. Therefore this `2.` point have quite a lot of consequences.

Reply:
How about supporting only `revision` in BuildRun? See why here https://github.com/redhat-developer/build/pull/218#discussion_r427170474 .

That way we still preserve the integrity in a reasonable way. ( source repository and branch remains intact ). 

I could add a little more confusion for future into the mix cc @zhangtbj . We could support triggering of BuildRuns from PRs which could imply arbitrary/forked repo + branch + revision. 

Therefore, the only bit of integrity that remains preserved then is: 
* Must be the same repo or branch or 
* a fork of the repo+branch.

Reply:
@sbose78 when you say `revision` what do you mean? 😅 At the moment in [code](https://github.com/redhat-developer/build/blob/master/pkg/apis/build/v1alpha1/gitsource.go#L15) is the `git branch` .

Reply:
(My bad, lack of coffee in the morning)
:D

Reply:
https://github.com/redhat-developer/build/issues/223

Reply:
when you define the revision in our git resource, it can be a commit, tag, branch, revision. Tekton can understand them well.

For my perspective, I think we don't need to allow end user to change the git url in the buildrun, because they can make sure it is a same build, it doesn't make sense if you define github.com/abc, but use a different github.com/efg in buildrun.

so the url of the fork branch is also different with the original url, i think we don't need to suppor it.

Reply:
+1 on @zhangtbj comment.

Reply:
Yeah!